### ISSUE TITLE
Babbage UI as modular components

### ISSUE THREAD

**Issue by pwalsh** [openspending/babbage.ui#16]
# Description

We want to guide the great work done in Babbage UI to date towards a more modular component architecture. 

Ideally, consumers of Babbage UI should be able to include components that they want into their code, such as, the Babbage Treemap, without any additional baggage. 

Of course, Consumers should also be able to simply bring the whole kitchen sink as a UI/Vis toolkit for apps that require multiple views on data provided by an instance of a Babbage API.

Additionally, modular Javascript often (usually) means writing with CommonJS or ES6 for a proper module implementation, and building *distributions* for use directly in the browser, or as importable libs in other CommonJS apps.

Babbage is currently written as a set of Angular 1. x directives, and is not currently usable as a modular "component library". This leads to a final reason why a move to modular components *can be* desirable: It moves towards conceptual alignment with the majority of modern Javascript frameworks, including Angular 2, and it is hoped that this will make a future port to Angular 2 less painful as we evolve the Babbage ecosystem.

Desired output for this work:

1. Use Webpack for building distributions
2. Publish the library as CommonJS and Browser distribution to npm, as well as the browser distribution to Bower
3. Each Visualisation, or View, as a distinct, self-contained component
4. Ability to import / require a single component. This would look something like the React-D3 bindings used [here](https://github.com/openspending/os-viewer/blob/master/src/scripts/components/views/Pie.js#L2), [here](https://github.com/openspending/os-viewer/blob/master/src/scripts/components/views/Pie.js#L15)
5. Optional, but strongly preferred by @pwalsh - Write `src` in ES6, which is transpiled to `lib` as ES5 - following the common pattern used these days for distributing ES6 code.

In any event, this can be done in stages, and the first, immediate goal would be the most minimal refactoring required to the existing code to be able to export the current directives for each visualisation as components.

Additionally, babbage.ui currently has some code that is not related to using babbage.ui as library of components, such as the S3 config in Grunt. My suggestion would be that we remove anything from babbage.ui that is not strictly required for creating the various visualisation widgets, and that this removed config and code could be pushed to a higher-level codebase that itself depends on babbage.ui as a component library.

References for the refactor:

* http://teropa.info/blog/2015/10/18/refactoring-angular-apps-to-components.html
* https://www.airpair.com/angularjs/posts/component-based-angularjs-directives

Work to be done in the [`feature/modern`](https://github.com/openspending/babbage.ui/tree/feature/modern) repository.
# Tasks

* [ ] TBD
* [ ] Basic test coverage

**Comment by pudo** [comment]
Brilliant, let's see how far we can get on this. A few points:

* I know @milafrerichs has an outstanding set of pull requests coming from his work on the RLP and Berlin budgets. My proposal would be to take these in before starting any major refactor, just to make sure his work is included - I think it'll lead in the right direction anyways. 

* Having only skimmed the first article (need more time!), I feel that most of the advice there is about moving everything to Angular 1.x directives. That shouldn't be much of a problem, since all of babbage.ui is directives already. Setting them up differently might be a good idea (e.g. I'm not sure how good an idea the query directive is, vs. a service).

* We can't adopt Angular 2 yet. It's not a thing yet, and doing that would defeat the whole purpose of making this library. Maybe in half a year, people will start using it, then this would make much more sense. In the same way, I propose we make this a really cool, idiomatic Angular 1.x library for now, and then move to 2.x when that makes sense. Trying too hard to play 2.x in a 1.x context seems a bit awkward.

* Please, let's make sure it keeps working :) I wouldn't want to break SpenDB just to achieve some architectural state of beauty.

**Comment by pwalsh** [comment]
@pudo 

1. I don't want to adopt Angular 2 at this stage. In making the case for components, I'm just pointing out that, in Angular terms, that is in the future anyway. I hope that is clear.
2. keeping it working: I want to be using this code in the coming weeks in OpenSpending - it needs to work :). In any event, we'll be on a different branch. Doing an output of babbage as per the current `dist/*` is trivial - nothing here changes the ability to output a browser-ready distribution as it currently is.
3. I'll personally ensure that all of the work that @milafrerichs does will get merged into the refactor.

**Comment by pwalsh** [comment]
Also I'm definitely going to be deferring some decisions to @kravets-levko as he very proficient in Angular and our main frontend developer, who will lead on this implementation. I'll make sure that we log issues/concerns/questions here so that all relevant parties can track the progress.

**Comment by pwalsh** [comment]
@pudo hey. So we've been working with babbage.ui for a while now, and are now merging in our component refactoring to the `feature/modern` branch, **not** to the master branch.

Our new code also deals with drilldowns, but in a different way to the PR from @milafrerichs (https://github.com/openspending/babbage.ui/pull/15) which was found to have some issues (not sure what they were but @borysyuk can elaborate if needed).

The API for each directive remains the same, except the `babbage` directive, as this is not used due to the componentization of the code. @kravets-levko worked a lot of this can can answer any specific questions on API compatibility.

Lastly, we are adding new components:

* GeoView (vis data on a map when we have a location dimension)
* BubbleTree

And more in the near future.

We are merging to `feature/modern` in order to be super clear that we do not effect the existing code. 

Then, when all the individual components get PR'd into `feature/modern`, we'll do a PR from there to `master` and note any and all API changes so that you can consider for support in SpenDB. We'll also help with specific changes where possible.

Is that all good? Are there specific questions I have not addressed? In the worst case, we can break out our branch to a new repo, as this work is necessary to make `babbage.ui` more generally usable, but I don't want to do that, so please keep in the loop so we can keep it workable against SpenDB too.

**Comment by pwalsh** [comment]
Oh, and also, the whole library will have tests - this is a pretty major enhancement to build out vis tools for babbage APIs :).

**Comment by borysyuk** [comment]
List parts of babbage.ui without refactoring:
Crosstab, Pager, Facts, Workspace (they have been used in pivot table)
and multi series Line/Bar charts

**Comment by pwalsh** [comment]
@borysyuk 

## Tasks

* [ ] Refactor & tests for multiseries line/bar component
* [ ] Refactor & tests for Facts component
* [ ] Refactor & tests for Pager component

**Comment by pwalsh** [comment]
@pudo 

We are essentially left with:

* Crosstab: We implemented using a different pivot table plugin. Therefore, not sure whether to port crosstab, or help this new PT work in spenDB. Any thoughts?
* Workspace and Panels: These are really the "application" part of babbage.ui, and not so useful as reusable components. Thinking out loud, the logic next step would be a distinct repo that is the "babbage.ui app", providing the interface you use in spendb, but using the new components. I'll help with this, of course, I just want to know what your thoughts are.

@borysyuk @akariv any comments on the advantage of the pivot table we implemented over crosstab here?

**Comment by pudo** [comment]
@pwalsh agreed that the app part might have to live somewhere else, but I do think it's really essential. A really flexible query tool is one of the things that OpenSpending never had and that made it a toy in the eyes of many more advanced users. For OSNext, is that something you want to have in the OSViewer package? Do you have mockups anywhere? 

In any case, I guess all I need is for there to be a tagged version of babbage.ui before you merge into master so that I can still deploy SpenDB until I get some time to upgrade the dependency over there. If that exists, please feel free to merge, and, I guess, commence user testing :)

**Comment by pwalsh** [comment]
Hey, just about the viewer mockups - I can go one better:

http://os-viewer.herokuapp.com

It loads a default data package alphabetically. To choose something (boost moldova is a decent example), then see the arrow on the top right of the workspace, which opens a search over all package names.

**Comment by pwalsh** [comment]
@akariv ok we need to get `feature/components` onto master. These are the steps I see *before* we do that:

- [ ] Get out the Babbage API wrapper into a standup lib (easier to use it for custom visualisations)
- [ ] Create a new branch that outputs the "spendb-compatible" version of the lib as per @pudo comment here https://github.com/openspending/babbage.ui/issues/16#issuecomment-204365606

Let's discuss/agree and then work out who does them (maybe I can).

**Comment by pwalsh** [comment]
@akariv can we do this? Actually, can @kravets-levko do it? I imagine for him it is not too much work, and it is related to the refactoring work he does now on the Viewer anyway. This is important so that other contributors can start contributing to the active branch.

**Comment by akariv** [comment]
@kravets-levko what do you think?

**Comment by kravets-levko** [comment]
I think it's really great idea to split `babbage.ui` into several more specific libraries. I can do it when I finish all current stuff on Viewer

**Comment by pudo** [comment]
Voice from the off: does it really need to be split? I feel like the degree to which OKI code is split into packages is somwhat disorienting and might have actually reached a threshold that will prevent adoption rather than foster it. [This scares me](http://frictionlessdata.io/tools/#catalog-of-tools-and-integrations
)...

**Comment by pwalsh** [comment]
@pudo point taken. But let's talk about this case, and not that list. The suggestion here is:

- A JS Client for Babbage API
- A JS library of views (treemap, bar, line, yada yada yada) that use the client

Seems quite reasonable to me (well, for obvious reasons :)), and quit a common pattern (why ship all the views with the API wrapper?). Such a separation of concerns would be quite consistent with JS packaging in general, and is not really OKI specific.

**Comment by pudo** [comment]
Sorry to generalise. Just out of interest: what would you use the Babbage API package for, if not tables, charts, etc.? Could the API-only package not be a second build target inside the same package?

**Comment by pwalsh** [comment]
@pudo yes, it could be a second build target in the same package, and TBH I'd be absolutely happy with that too, so if @akariv and @kravets-levko prefer that too, then right on :).
